src/core/url-validation.ts

```diff
@@ -0,0 +1,652 @@
+        return false
+      }
+
+      // Block protocol-relative URLs (//host)
+      if (data.startsWith('//')) {
+        return false
+      }
+
+      const parsed = safeParseUrl(data)
+      if (!parsed) {
+        return false
+      }
+
+      const { url: urlObj, wasRelative } = parsed
+
+      // Check protocol - only allow http, https, and relative protocols
+      if (urlObj.protocol !== 'http:' && urlObj.protocol !== 'https:' && urlObj.protocol !== '') {
+        return false
+      }
+
+      // Check for credentials and backslashes
+      if (hasCredentials(urlObj) || hasBackslash(data) || hasBackslash(urlObj.href)) {
+        return false
+      }
+
+      // Only check host/port/SSRF for absolute URLs
+      if (!wasRelative) {
+        // Check for disallowed ports
+        if (isDisallowedPort(urlObj)) {
+          return false
+        }
+
+        // Check for private hosts (SSRF protection)
+        // Note: This is static validation. Complete SSRF protection requires DNS resolution +
+        // IP validation on the server before opening sockets
+        if (urlObj.hostname && isPrivateHost(urlObj.hostname)) {
+          return false
+        }
+      }
+
+      // Check path for directory traversal and malicious paths
+      if (!isSafePath(urlObj.pathname)) {
+        return false
+      }
+
+      // Check path length limit (2KB)
+      if (urlObj.pathname.length > 2048) {
+        return false
+      }
+
+      // Check search parameters for malicious content and limits
+      const paramEntries = Array.from(urlObj.searchParams.entries())
+      if (paramEntries.length > 50) {
+        return false // Too many parameters
+      }
+
+      // Check total query string length (2KB limit)
+      if (queryLength(urlObj) > 2048) {
+        return false
+      }
+
+      for (const [key, value] of paramEntries) {
+        // Reject empty keys
+        if (key.length === 0) {
+          return false
+        }
+
+        if (!isSafeString(key) || !isSafeString(value)) {
+          return false
+        }
+
+        // Check parameter size limits
+        if (key.length > 128 || value.length > 1024) {
+          return false
+        }
+      }
+
+      // Check fragment for control characters
+      if (urlObj.hash && hasControlOrBidi(urlObj.hash)) {
+        return false
+      }
+
+      return true
+    },
+    errors: false
+  }
+
+  export const schema = {
+    type: 'string',
+    _isSafeUrl: true
+  } as JSONSchema<string>
+
+  export const validate: ValidateFunction<string> = generateLazyValidator(schema, [_isSafeUrl])
+
+  /**
+   * Safely builds a URL with query parameters, automatically encoding all values
+   * @param basePath - The base path without query parameters
+   * @param params - Object containing query parameters
+   * @returns A properly encoded URL string
+   * @throws Error if malicious content is detected
+   */
+  export function buildUrlWithParams(basePath: string, params: QueryParams): string {
```

Comment:
What if, instead of having an URL builder, which the URL class already has (and does it correctly), we have a URL instance checker? We can even convert the URL to string and use the string checker that was defined before.